### ISSUE TITLE
GEOS-Chem 14.2.0 development branch crashing with integration errors

### ISSUE THREAD

**Issue by msulprizio** [geoschem/geos-chem#1768]
### Name and Institution (Required)

Name: Melissa Sulprizio
Institution: Harvard/GCST

### Confirm you have reviewed the following documentation

- [x] [Support guidelines](https://geos-chem.readthedocs.io/en/stable/reference/SUPPORT.html)
- [x] [User manuals](https://geos-chem.readthedocs.io/en/stable/geos-chem-shared-docs/supplemental-guides/related-docs.html)
- [x] [Debugging GEOS-Chem and HEMCO errors](https://geos-chem.readthedocs.io/en/stable/geos-chem-shared-docs/supplemental-guides/debug-guide.html)
- [x] [Current and past Github issues](https://github.com/geoschem/geos-chem/issues)

### Description of your issue or question

When running a 4x5 GEOS-FP simulation for 2016 with GEOS-Chem 14.2.0-alpha.8, Nadia Colombi's runs crash with an integration error three months into the simulation with:

```
---> DATE: 2016/03/05  UTC: 15:20
 Forced exit from Rosenbrock due to the following error:
 --> Step size too small: T + 10*H = T or H < Roundoff
 T=   4.8000000000000068E-018 and H=   4.8000000000000068E-018
 ### INTEGRATE RETURNED ERROR AT:           54          44          53
 Forced exit from Rosenbrock due to the following error:
 --> Step size too small: T + 10*H = T or H < Roundoff
 T=   4.0000000000000041E-018 and H=   4.0000000000000041E-018
## INTEGRATE FAILED TWICE !!! 
 #############################################################################################################################################################################################################################################
 ### KPP DEBUG OUTPUT!
 ### Species concentrations at problem box           54          44          53
 #############################################################################################################################################################################################################################################
 ###    1.5330768753631170E-029 CH2I2
 ###    9.7494583413162528E-025 CH2IBr
 ###    2.3087058718844786E-023 CH2ICl
 ###    37627.245756688040      AERI
 ###    12612.597852921273      AONITA
 ...
```

The full log file may be viewed [here](https://github.com/geoschem/geos-chem/files/11295403/GC.log). 

This version includes several fixes and updates for chemistry and sea salt alkalynity from Becky Alexander, SSA debromination on, and nitrate photolysis on. See the [14.2.0 milestone](https://github.com/geoschem/geos-chem/milestone/36?closed=1) for the complete list of updates. The simulation was initialized with a Jan 2016 restart file from the 14.0.0 10-year benchmark.

We are still investigating.

Also tagging @yantosca @ncolombi @beckyalexander @viral211

**Comment by msulprizio** [comment]
Some debug testing notes:
- Running with debug turned on did not catch any other issues (div-by-zero, array-out-of-bounds, etc).
- Switching to MERRA-2 did not resolve the issue. This was to test if the error was caused by a bad GEOS-FP met field on 03/05/2016.
- Reverting relative tolerance (RTOL) in fullchem_mod.F90 back to 0.5e-2_dp (from the updated 0.5e-3_dp value) didn't resolve the issue. The run crashed earlier in the day.
- Running with default emissions for benchmark simulations did not resolve the issue. Nadia's runs used OFFLINE_BIOGENICVOC, OFFLINE_SOILNOX, and nonstandard inventories KORUS and MEIC, but these do not seem to be the problem.

The above tests were run with restart files spun up from Nadia's original run. I am also spinning up my own restart files to see if the issue is cumulative (e.g. species concentrations getting extremely small with time, eventually causing the model to crash).

**Comment by viral211** [comment]
From the log file, the problem seems to be with the rates of these iodine reactions:
 ```
  5.3118017604483310E+033 HOI --> ISALC + LOx
  2.8298654872890720E+033 IONO --> ISALC + LOx
  5.8315631411491239E+033 IONO2 --> ISALC + 2 LOx
```

**Comment by viral211** [comment]
We should cap the fraction of alkaline sea salt (`H%f_Alk_SSA` and `H%f_Alk_SSC`) to 1. `H%f_Alk_SSC` has quite large values in the stratosphere (not sure why).

**Comment by beckyalexander** [comment]
I've been wondering if we should perform reactions on sea salt aerosol in the troposphere only. It is not currently written this way in the code. This would be an easy fix.

Also, I noticed recently that the reactions should look like the following so that they are balanced:
IONO --> ISALC + HNO2
IONO2 --> ISALC + HNO3
A similar change should be made for the reactions producing ISALA. I made this change in my working version and the changes for O3, NOx, HNO2, NIT, and HNO3 are negligible.

[zonal_mean_INbalance_HNO2_O3.pdf](https://github.com/geoschem/geos-chem/files/11297836/zonal_mean_INbalance_HNO2_O3.pdf)
[INbalance_NO_NO2.pdf](https://github.com/geoschem/geos-chem/files/11297837/INbalance_NO_NO2.pdf)
[INbalance_HNO3_NIT.pdf](https://github.com/geoschem/geos-chem/files/11297838/INbalance_HNO3_NIT.pdf)

**Comment by yantosca** [comment]
> We should cap the fraction of alkaline sea salt (`H%f_Alk_SSA` and `H%f_Alk_SSC`) to 1. `H%f_Alk_SSC` has quite large values in the stratosphere (not sure why).

@viral211: This might be a division issue where the denominator gets very small and blows up the result.  We do have a trap to set it to 0 if the division can't be done, but in the case where SALA or SALC concentrations get very small, we don't have a trap to prevent it from blowing up.  See:

https://github.com/geoschem/geos-chem/blob/5da602bc7d274cd6d96b33ac5f21699974d78331/KPP/fullchem/fullchem_HetStateFuncs.F90#L156-L163

We can put a trap in to cap the values at 1 and see if this lets us proceed w/o problems.

**Comment by yantosca** [comment]
This code should work:
```F90
H%f_Alk_SSA     = SafeDiv( State_Chm%Species(id_SALAAL)%Conc(I,J,L),     & 
                           State_Chm%Species(id_SALA  )%Conc(I,J,L),     & 
                           0.0_dp                                       ) 
H%f_Alk_SSA     = MIN( H%f_Alk_SSA, 1.0_dp )
H%f_Acid_SSA    = 1.0_dp - H%f_Alk_SSA 
H%f_Alk_SSC     = SafeDiv( State_Chm%Species(id_SALCAL)%Conc(I,J,L),     & 
                           State_Chm%Species(id_SALC  )%Conc(I,J,L),     & 
                            0.0_dp                                       ) 
H%f_Alk_SSC     = MIN( H%f_Alk_SSC, 1.0_dp )
H%f_Acid_SSC    = 1.0_dp - H%f_Alk_SSC
```

**Comment by yantosca** [comment]
> I've been wondering if we should perform reactions on sea salt aerosol in the troposphere only. It is not currently written this way in the code. This would be an easy fix.

@viral211 @beckyalexander: I can implement this.  I just wanted to confirm which rate-law functions in which we should return k=0 if we are int the stratosphere.  It looks like we'd want to do this for these iodine reactions, but are there any others?

```F90    
HI = ISALA :                   IuptkBySALA1stOrd( SR_MW(ind_HI), 0.10_dp, State_Het );                      {2017/09/22; Sherwen2016b;TS,JAS,SDE}
HI = ISALC :                   IuptkBySALC1stOrd( SR_MW(ind_HI), 0.10_dp, State_Het );                      {2017/09/22; Sherwen2016b;TS,JAS,SDE}
HOI = ISALA :                  IuptkByAlkSALA1stOrd( SR_MW(ind_HOI), 0.01_dp, State_Het );                  {2023/01/24; Restored deleted rxn:BA}
HOI = ISALC :                  IuptkByAlkSALC1stOrd( SR_MW(ind_HOI), 0.01_dp, State_Het );                  {2022/12/02; Restored deleted rxn:BA}
I2O2 = 2.000AERI :             IuptkBySulf1stOrd( SR_MW(ind_I2O2), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
I2O2 = 2.000ISALA :            IuptkBySALA1stOrd( SR_MW(ind_I2O2), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
I2O2 = 2.000ISALC :            IuptkBySALC1stOrd( SR_MW(ind_I2O2), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
I2O3 = 2.000AERI :             IuptkBySulf1stOrd( SR_MW(ind_I2O3), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
I2O3 = 2.000ISALA :            IuptkBySALA1stOrd( SR_MW(ind_I2O3), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
I2O3 = 2.000ISALC :            IuptkBySALC1stOrd( SR_MW(ind_I2O3), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
I2O4 = 2.000ISALA :            IuptkBySALA1stOrd( SR_MW(ind_I2O4), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
I2O4 = 2.000ISALC :            IuptkBySALC1stOrd( SR_MW(ind_I2O4), 0.02_dp, State_Het );                    {2017/09/22; Sherwen2016b;TS,JAS,SDE}
IONO = ISALA :                 IuptkByAlkSALA1stOrd( SR_MW(ind_IONO), 0.02_dp, State_Het );                 {2022/12/02; Restored deleted rxn:BA}
IONO = ISALC :                 IuptkByAlkSALC1stOrd( SR_MW(ind_IONO), 0.02_dp, State_Het );                 {2022/12/02; Restored deleted rxn:BA}
IONO2 = ISALA :                IuptkByAlkSALA1stOrd( SR_MW(ind_IONO2), 0.01_dp, State_Het );                {2022/12/02; Restored deleted rxn:BA}
IONO2 = ISALC :                IuptkByAlkSALC1stOrd( SR_MW(ind_IONO2), 0.01_dp, State_Het );                {2022/12/02; Restored deleted rxn:BA}
IONO + BrSALA = IBr + HNO2 :   IbrkdnByAcidBrSALA( SR_MW(ind_IONO), C(ind_IONO), 0.02_dp, State_Het );      {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
IONO + BrSALC = IBr + HNO2 :   IbrkdnByAcidBrSALC( SR_MW(ind_IONO), C(ind_IONO), 0.02_dp, State_Het );      {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
IONO + SALACL = ICl + HNO2 :   IbrkdnByAcidSALACl( SR_MW(ind_IONO), C(ind_IONO), 0.02_dp, State_Het );      {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
IONO + SALCCL = ICl + HNO2 :   IbrkdnByAcidSALCCl( SR_MW(ind_IONO), C(ind_IONO), 0.02_dp, State_Het );      {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
IONO2 + BrSALA = IBr + HNO3 :  IbrkdnByAcidBrSALA( SR_MW(ind_IONO2), C(ind_IONO2), 0.01_dp, State_Het );    {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
IONO2 + BrSALC = IBr + HNO3 :  IbrkdnByAcidBrSALC( SR_MW(ind_IONO2), C(ind_IONO2), 0.01_dp, State_Het );    {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
IONO2 + SALACL = ICl + HNO3 :  IbrkdnByAcidSALACl( SR_MW(ind_IONO2), C(ind_IONO2), 0.01_dp, State_Het );    {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
IONO2 + SALCCL = ICl + HNO3 :  IbrkdnByAcidSALCCl( SR_MW(ind_IONO2), C(ind_IONO2), 0.01_dp, State_Het );    {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
HOI + BrSALA = IBr :           IbrkdnByAcidBrSALA( SR_MW(ind_HOI), C(ind_HOI), 0.01_dp, State_Het );        {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
HOI + BrSALC = IBr :           IbrkdnByAcidBrSALC( SR_MW(ind_HOI), C(ind_HOI), 0.01_dp, State_Het );        {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
HOI + SALACL = ICl :           IbrkdnByAcidSALACl( SR_MW(ind_HOI), C(ind_HOI), 0.01_dp, State_Het );        {2017/09/22; Sherwen2017;TS,JAS,SDE,XW}
HOI + SALCCL = ICl :           IbrkdnByAcidSALCCl( SR_MW(ind_HOI), C(ind_HOI), 0.01_dp, State_Het );   
```
Let me know and I'll code it up.

**Comment by beckyalexander** [comment]
@yantosca: Yes, the iodine reactions that you suggest and probably also the Br and Cl reactions. @viral211, what do you think? 

N2O5 + SALACL = ClNO2 + HNO3 : N2O5uptkBySALACl( State_Het );                                               {2018/01/19; Sherwen2017;TS,JAS,SDE,XW}
N2O5 + SALCCL = ClNO2 + HNO3 : N2O5uptkBySALCCl( State_Het );                                               {2018/01/19; Sherwen2017;TS,JAS,SDE,XW}
OH + SALACL = 0.500Cl2 :       OHuptkBySALACl( State_Het );                                                 {2018/03/12; XW}
OH + SALCCL = 0.500Cl2 :       OHuptkBySALCCl( State_Het );                                                 {2018/03/12; XW}
ClNO3 + BrSALA = BrCl + HNO3 : ClNO3uptkByBrSALA( State_Het );                                              {2017/09/22; Sherwen2016b;TS,JAS,SDE}
ClNO3 + BrSALC = BrCl + HNO3 : ClNO3uptkByBrSALC( State_Het );                                              {2017/09/22; Sherwen2016b;TS,JAS,SDE}
ClNO3 + SALACL = Cl2 + HNO3 :  ClNO3uptkBySALACL( State_Het );                                              {2018/01/22; XW}
ClNO3 + SALCCL = Cl2 + HNO3 :  ClNO3uptkBySALCCL( State_Het );                                              {2018/01/22; XW}
ClNO2 + SALACL = Cl2 + HNO2 :  ClNO2uptkBySALACL( State_Het );                                              {2018/01/22; XW}
ClNO2 + SALCCL = Cl2 + HNO2 :  ClNO2uptkBySALCCL( State_Het );                                              {2018/01/22; XW}
ClNO2 + BrSALA = BrCl + HNO2 : ClNO2uptkByBrSALA( State_Het );                                              {2018/01/22; XW}
ClNO2 + BrSALC = BrCl + HNO2 : ClNO2uptkByBrSALC( State_Het );                                              {2018/01/22; XW}
HOCl + SALACL = Cl2 + H2O :    HOClUptkBySALACL( State_Het );                                               {2018/01/22; XW}
HOCl + SALCCL = Cl2 + H2O :    HOClUptkBySALCCL( State_Het );                                               {2018/01/22; XW}
HOCl + BrSALA = BrCl + H2O :   HOClUptkByBrSALA( State_Het );                                               {2023/03/31; BA}
HOCl + BrSALC = BrCl + H2O :   HOClUptkByBrSALC( State_Het );                                               {2023/03/31; BA}
HOBr + SALACL = BrCl + H2O :   HOBrUptkBySALACL( State_Het );                                               {2018/01/22; Sherwen2017;TS,JAS,SDE;XW}
HOBr + SALCCL = BrCl + H2O :   HOBrUptkBySALCCL( State_Het );                                               {2018/01/22; Sherwen2017;TS,JAS,SDE,XW}
HOBr + BrSALA = Br2 :          HOBrUptkByBrSALA( State_Het );                                               {2017/09/22; Sherwen2017;TS,JAS,SDE}
HOBr + BrSALC = Br2 :          HOBrUptkByBrSALC( State_Het );                                               {2017/09/22; Sherwen2017;TS,JAS,SDE}
O3 + BrSALA = HOBr :           O3uptkByBrSALA( State_Het );                                                 {2017/09/22; Sherwen2016b;TS,JAS,SDE}
O3 + BrSALC = HOBr :           O3uptkByBrSALC( State_Het );                                                 {2017/09/22; Sherwen2016b;TS,JAS,SDE}HBr = BrSALA :                 
HBr = BrSALA :                 HBrUptkBySALA( State_Het );                                                  {2017/09/22; Sherwen2016b;TS,JAS,SDE}
HBr = BrSALC :                 HBrUptkBySALC( State_Het );                                                  {2017/09/22; Sherwen2016b;TS,JAS,SDE}

**Comment by viral211** [comment]
I agree with @beckyalexander that we don't need to have these reactions in the stratosphere. The only one from the list I think we should retain in the stratosphere is: `I2O3 = 2.000AERI : IuptkBySulf1stOrd( SR_MW(ind_I2O3), 0.02_dp, State_Het );` since it takes place on strat. sulfate aerosols too.

**Comment by yantosca** [comment]
Thanks @beckyalexander and @viral211!  I've just pushed commit 1a7c81b6e013b58c87b2bb97228f71459fa5149f where I exit from several rate law routines with k = 0 if we are in the stratosphere.  I didn't touch `IuptkBySulf1stOrd`.  Also a few of those rate law functions already had an IF block to not do certain computations in the stratosphere, so I didn't touch those.

I will create a PR for this shortly.

**Comment by yantosca** [comment]
We can close out this issue since #1779 has been merged into the 14.2.0 development stream.  1-year simulations have proceeded without error since implementing this fix.
